flutter-winit/src/window.rs

```diff
@@ -0,0 +1,537 @@
+                                        });
+                                    });
+                                }
+                                ElementState::Released => {
+                                    engine.with_plugin_mut(|keyevent: &mut KeyEventPlugin| {
+                                        keyevent.key_action(KeyAction {
+                                            toolkit: "glfw".to_string(),
+                                            key_code: raw_key as _,
+                                            scan_code: scancode as _,
+                                            modifiers: raw_modifiers as _,
+                                            keymap: "linux".to_string(),
+                                            _type: KeyActionType::Keyup,
+                                        });
+                                    });
+                                }
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                Event::LoopDestroyed => {
+                    engine.shutdown();
+                }
+                _ => {
+                    if close.load(Ordering::Relaxed) {
+                        *control_flow = ControlFlow::Exit;
+                        return;
+                    }
+
+                    let next_task_time = engine.execute_platform_tasks();
+
+                    if let Some(next_task_time) = next_task_time {
+                        *control_flow = ControlFlow::WaitUntil(next_task_time)
+                    } else {
+                        *control_flow = ControlFlow::Wait
+                    }
+                }
+            });
+    }
+
+    pub fn wake_platform_thread(&self) {
+        self.engine_handler.wake_platform_thread();
+    }
+}
+
+fn resize(engine: &FlutterEngine, context: &Arc<Mutex<Context>>) {
+    let mut context = context.lock();
+    let dpi = context.hidpi_factor();
+    let size = context.size().to_physical(dpi);
+    log::trace!(
+        "resize width: {} height: {} scale {}",
+        size.width,
+        size.height,
+        dpi
+    );
+    context.resize(size);
+    engine.send_window_metrics_event(size.width as usize, size.height as usize, dpi);
+}
+
+fn text_input_shortcuts(engine: &FlutterEngine, key: VirtualKeyCode) {
+    engine.with_plugin_mut(|text_input: &mut TextInputPlugin| match key {
+        VirtualKeyCode::Return => {
+            text_input.with_state(|state| {
+                state.add_characters(&"\n");
+            });
+            text_input.notify_changes();
+        }
+        VirtualKeyCode::Back => {
+            text_input.with_state(|state| {
+                state.backspace();
+            });
+            text_input.notify_changes();
+        }
+        _ => {}
+    });
+}
+
+fn raw_key(key: Option<VirtualKeyCode>) -> u32 {
+    if let Some(key) = key {
+        use VirtualKeyCode as Key;
+        match key {
+            Key::A => 65,
```

Comment:
Where did this mapping come from? It might be worth adding a comment so that it can be easily maintained in the future.

It might be worth switching to looking up in an array, as I think matches on sparse values is linear.

Reply:
I considered that but then decided against it as it would rely on implementation details of rustc (it's not #[repr(C)] so there is no guarantees) and the VirtualKeyCode enum.